[Console] Monaco migration: send request and copy as curl buttons

yuliacech opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/178990

This PR adds the actions buttons to the monaco editor and the functionality for the buttons to send a request and copy the request as a curl command. When the cursor or user selection doesn't overlap or contain any requests on the selected line, the actions buttons are hidden. When the cursor or selection changes, the buttons are displayed on the 1st line of the 1st selected request. Several requests can be sent at once. Only the 1st request is copied as a curl command.

There are also some minor UI and copy changes (see screenshots below) as suggested by @MichaelMarcialis in the review.  

### Screenshot 

#### Before

<img width="882" alt="Screenshot 2024-04-15 at 15 16 59" src="https://github.com/elastic/kibana/assets/6585477/8fa95a5b-51c6-4220-8837-38adc4696602">
<img width="296" alt="Screenshot 2024-04-15 at 15 17 07" src="https://github.com/elastic/kibana/assets/6585477/6c99a4db-f7a6-4872-a24a-cddb56e0ec3d">

#### After
<img width="916" alt="Screenshot 2024-04-15 at 15 18 08" src="https://github.com/elastic/kibana/assets/6585477/1ae0b161-c731-42e1-90af-69468e9f0905">
<img width="367" alt="Screenshot 2024-04-15 at 15 18 14" src="https://github.com/elastic/kibana/assets/6585477/bb8818bd-38d4-4ec3-8bf4-30e24afd3664">


### How to test
- Check that the actions buttons are not displayed when there are no requests in the input
- Check that the actions buttons are not displayed when no requests are selected by the cursor or the selection
- Check that the selected requests are highlighted
- Check that the buttons are displayed on the 1st line of the 1st selected request
- Check that the position of the buttons is updated when the editor is scrolled up or down

### Follow up issues 
- The functionality for the button to open a documentation for the 1st selected request 
- The functionality for the button to auto-indent input
- The input highlighting is temporarily reset when sending a request
- Sent requests need to be saved to history
- After sending a request, the autocomplete polling needs to be restarted
- Add more unit tests

yuliacech commented on packages/kbn-monaco/src/console/console_errors_provider.ts:
```diff
@@ -0,0 +1,56 @@
+import { CONSOLE_LANG_ID } from './constants';
+import { monaco } from '../monaco_imports';
+
+export const setupConsoleErrorsProvider = (workerProxyService: ConsoleWorkerProxyService) => {
```
this file is copied from the previously used file `packages/kbn-monaco/src/ace_migration/setup_worker.ts`

yuliacech commented on packages/kbn-monaco/src/console/console_worker_proxy.ts:
```diff
@@ -0,0 +1,32 @@
+import { CONSOLE_LANG_ID } from './constants';
+import { ConsoleParserResult, ConsoleWorkerDefinition } from './types';
+
+export class ConsoleWorkerProxyService {
```
this file is copied from the previously used file `packages/kbn-monaco/src/ace_migration/worker_proxy.ts`

yuliacech commented:
/ci

yuliacech commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

MichaelMarcialis commented on src/plugins/console/public/application/containers/editor/monaco/monaco_editor.tsx:
```diff
@@ -39,6 +62,42 @@ export const MonacoEditor = ({ initialTextValue }: EditorProps) => {
+          </EuiToolTip>
+        </EuiFlexItem>
+        <EuiFlexItem>
+          <ConsoleMenu
+            getCurl={getCurlCallback}
+            getDocumentation={() => {
+              return Promise.resolve(null);
+            }}
+            autoIndent={() => {}}
+            notifications={notifications}
+          />
```
Is the current "Open documentation" action intend to be contextual to the currently highlighted command? If not, would both "Open documentation" and "Auto indent" actions be better served as global actions rather than individual command actions?

MichaelMarcialis reviewed (commented):
Thanks for the ping, @yuliacech! I'm coming in a bit sideways, so I've left you a handful of comments/questions for my initial review. Let me know if there's anything than needs additional clarity.

ElenaStoeva reviewed (commented):
Amazing job @yuliacech! I tested locally and everything implemented so far works well (apart from the highlighting issue but this is not related to this PR). I left a few comments  where I thought documentation or unit tests could be added.

yuliacech commented:
Thanks a lot for your review, @MichaelMarcialis! I implemented your suggestions for the buttons, except for usage of `EuiButtonIcon`: it renders the icons with some padding that messes up the alignment of the buttons on the line. 
<img width="717" alt="Screenshot 2024-04-15 at 15 44 02" src="https://github.com/elastic/kibana/assets/6585477/58ab22b5-5c6a-40fc-905c-f412bcc144af">
<img width="331" alt="Screenshot 2024-04-15 at 15 44 11" src="https://github.com/elastic/kibana/assets/6585477/3e471fe1-910a-4091-ab44-54c70318a26f">
So I kept `EuiLink` with a nested icon for now, since it results in the same html structure as `EuiButtonIcon` anyways (button with an svg). Let me know if you think we should address this in a follow up issue. 
To answer your question about documentation and auto-indent buttons: this functionality is not yet implemented, we have https://github.com/elastic/kibana/issues/180209 and https://github.com/elastic/kibana/issues/180213 to to track this work. Both actions apply to the "current" request, so the browser will open a new tab with the docs for the specific request and also indentations are applied only to the selected request.

yuliacech commented:
Thanks a lot for your review, @ElenaStoeva! 
I added comments explaining what some of the classes are used for, let me know if that makes sense or if more information could be useful there. Also added more unit tests as you suggested.

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @yuliacech! Looks great.

> I implemented your suggestions for the buttons, except for usage of EuiButtonIcon: it renders the icons with some padding that messes up the alignment of the buttons on the line.

>So I kept EuiLink with a nested icon for now, since it results in the same html structure as EuiButtonIcon anyways (button with an svg). Let me know if you think we should address this in a follow up issue.

Yeah, this might be worthwhile to revisit in a follow-up issue. While the resulting HTML may be the same or similar, the button component styling is a bit more robust in regard to the hover/focus/active states, while the link component for icon-only content is a bit lacking. Things that jump to mind that may remedy the alignment issue you reference could be to 1) increase the console editor's line height, 2) override the EUI button styles to be slightly smaller, or 3) some combination of the two.

> To answer your question about documentation and auto-indent buttons: this functionality is not yet implemented, we have https://github.com/elastic/kibana/issues/180209 and https://github.com/elastic/kibana/issues/180213 to to track this work. Both actions apply to the "current" request, so the browser will open a new tab with the docs for the specific request and also indentations are applied only to the selected request.

Gotcha. In that case then, the only concern I'd mention currently is that the documentation link doesn't work in some situations.  For example, incomplete or invalid commands appear to result in nothing happening when attempting to view the documentation. Should we disable or hide the documentation link in those conditional situations?

Otherwise, this looks good from my perspective. Approving now so I don't hold you up further.

Dosant reviewed (approved):
Amazing work 👏 
Approving to unblock, haven't tested

ElenaStoeva reviewed (approved):
Thanks for addressing my feedback @yuliacech!

yuliacech commented:
Opened https://github.com/elastic/kibana/issues/180911 to follow up on the action buttons
